### PR TITLE
Create streams directory

### PR DESCRIPTION
**Pull request by Brett208**
How to we want to handle including the Stream directory in other files?

I was thinking about leaving a StreamWriter.h and StreamReader.h header. These would each bring in 4 class StreamWriter, SeekableStreamWriter, FileStreamWriter, and MemoryStreamWriter.

To do this, I guess I need to create a StreamWriterBase.h and SeekableStreamWriter.h header. Otherwise I would have to place the FileStreamWriter and MemoryStreamWriter includes below the StreamWriter and SeekableStreamWriter definitions in StreamWriter.h. This seems wrong. Check the current commit on this branch to see this in action.

We could force the end user to bring in each individually. IE if they want to use FileStreamWriter, they would include FileStreamWriter.h. This seems like a lot of includes though because they probably are going to want the SeekableStreamWriter.h as well if not other stream classes.

Thanks,
Brett

### CODE REVIEW

File: OP2Utility.h [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #pragma once
 
 #include "StringHelper.h"
-#include "StreamReader.h"
+#include "Streams/StreamReader.h"
```
Comment (DanRStevens) [inline OP2Utility.h]: What is the purpose of OP2Utility.h? Is this a generic import header to bring in everything the library has to offer, or does it have more limited project specific usage? If it's a generic import header, then it should probably import everything. If not, I assume it's correct as is.

File: Streams/MemoryStreamWriter.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include "StreamWriter.h"
```
Comment (DanRStevens) [inline Streams/MemoryStreamWriter.h]: This should import SeekableStreamWriter, the direct parent. The SeekableStreamWriter class will need to import StreamWriter, so we get that header indirectly.

File: Streams/StreamWriter.cpp [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#include "StreamWriter.h"
+
+StreamWriter::~StreamWriter() { }
```
Comment (DanRStevens) [inline Streams/StreamWriter.cpp]: StreamWriter is an interface, so we could reasonably omit a .cpp file. It's possible to allow for an inline empty destructor in the header file even though it's declared pure virtual. The pure virtual means that method doesn't require an implementation, and marks the class as abstract. It does not mean the function does not have an implementation. A rather subtle distinction.

File: Streams/StreamWriter.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	
+	// Seek to absolute position, given as offset from beginning of stream
+	virtual void Seek(uint64_t offset) = 0;
+
+	// Seek by a relative amount, given as offset from current position
+	virtual void SeekRelative(int64_t offset) = 0;
+};
+
+#include "FileStreamWriter.h"
+#include "MemoryStreamWriter.h"
```
Comment (DanRStevens) [inline Streams/StreamWriter.h]: We should not be including the headers for derived objects. The inclusion should go the other direction.

Comment (Brett208) [inline Streams/StreamWriter.h]: Just realized the GitHub flags when no newline is detected at the end of a file. Kind of neat feature.

**Comment by Brett208** [comment]
Thanks for the thoughts on how to structure the headers.

It didn't occur to me that FileStreamWriter would automatically bring SeekableStreamWriter and StreamWriter into a project when referenced. So you could create a FileStreamWriter and reference it as a SeekableStreamWriter later in the file without adding a new include. 

For now I'll create individual headers for each stream class. If we feel like it adds too many headers into the rest of the files referencing streams, later we can add some sort of convenience header(s).

**Comment by Brett208** [comment]
Should resolve issue #43 

Merge conflicts are too complex to resolve using the GitHub web editor. I'll resolve them locally when ready to merge.

I'm not sure how I feel about the complex file structure in the includes. For example: '#include "../Streams/MemoryStreamReader.h"'. On one hand it is nice to know exactly where the header is, but it also sort of clutters the includes.

File: Streams/StreamWriter.h (hunk 4 above)
Comment (DanRStevens) [reply Streams/StreamWriter.h]: Yes, handy indeed. Some editors also have settings to warn about missing newlines, or to automatically insert newlines at the end of files. I don't much like the automatic edits myself, but I do like the show whitespace setting so I can see when something is off. It felt a bit odd and ugly at first, but I quickly came to love it.

**Comment by Brett208** [comment]
Sounds good on the includes. Lets leave as is for now then.

**Pull request merged**